[FEATURE] cast datasets pil objects to base64 in SDK

burtenshaw opened the pull request:
# Description
<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
This PR introduces a utility into the SDK so that it automatically casts PIL objects to base64 based on the datasets `Features`

**Type of change**
<!--  Please delete options that are not relevant. Remember to title the PR according to the type of change  -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Refactor (change restructuring the codebase without changing functionality)
- Improvement (change adding some improvement to an existing functionality)
- Documentation update

**How Has This Been Tested**
<!--  Please add some reference about how your feature has been tested.  -->

**Checklist**
<!--  Please go over the list and make sure you've taken everything into account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

burtenshaw committed:
feat: add media function to encode PIL as base64

burtenshaw committed:
feat: add casting methods to datasets records loaders

burtenshaw committed:
chore: add pil to dependencies

burtenshaw committed:
test: add tests for media util

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

jfcalvo commented on argilla/src/argilla/media.py:
```diff
@@ -0,0 +1,50 @@
+
+    try:
+        buffered = io.BytesIO()
+        image_object.save(buffered, format="PNG")
```
What's the motivation behind transforming it to PNG? I believe we should try not to process the image and use it as it is.

As a possible alternative we should try to get the mime-type (and use it inside the data-url) and if it's not supported by the backend raise an error.

jfcalvo replied:
You can get the list of supported mime types here: https://github.com/argilla-io/argilla/pull/5279/files#diff-05ee0287277b4f0bb73f7d03262800ff0cfb195ab90e9f11e29105eb88b51dcfR33

frascuchon commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -58,11 +60,65 @@ def _record_dicts_from_datasets(dataset: HFDataset) -> List[Dict[str, Union[str,
         return record_dicts
+
+    @staticmethod
+    def _get_media_features(dataset: "HFDataset") -> List[str]:
```
maybe rename it to `get_image_features` until we support other media types?

frascuchon commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -58,11 +60,65 @@ def _record_dicts_from_datasets(dataset: HFDataset) -> List[Dict[str, Union[str,
+        for name, feature in dataset.features.items():
+            if isinstance(feature, Image):
+                media_features.append(name)
+        return media_features
```
list-comprehension approach. Feel free to apply this change.
```suggestion
        media_features = [
           name
           for name, feature in dataset.features.items()
           isinstance(feature, Image)
        ]
        return media_features
```

frascuchon commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -58,11 +60,65 @@ def _record_dicts_from_datasets(dataset: HFDataset) -> List[Dict[str, Union[str,
+                original_column_name=f"column_{unique_identifier}", new_column_name=column
+            )
+
+        return hf_dataset
```
Since datasets supports modifying column values directly, what about this approach? 
```suggestion
        def batch_fn(batch, columns: List[str]):
            return {
                column: [pil_to_data_uri(sample[column]) for sample in batch]
                for column in columns 
            }

        return hf_dataset.map(
            function=batch_fn,
            batched=True,
            fn_kwargs={"columns": columns},
        )
```

burtenshaw replied:
@frascuchon I tested this out and an error is raise when we iterate over. I think this is because the dataset's `Features` need to match the samples.
 
I worked on this for a while trying a few approaches like `Dataset.cast_column` and found that this was the best I could get. 

Happy to try out more though.

frascuchon replied:
I thought we could apply the image conversion for all the image columns at once. That was my motivation, but the code is fine as is.

davidberenstein1957 commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -58,11 +60,65 @@ def _record_dicts_from_datasets(dataset: HFDataset) -> List[Dict[str, Union[str,
+            bool: True if the Hugging Face dataset contains image features, False otherwise.
+        """
+        media_features = []
+        for name, feature in dataset.features.items():
```
@burtenshaw I am not entirely sure how the `Image` feature for datasets works exactly. After [loading an image dataset](https://huggingface.co/docs/datasets/en/image_load#local-files), I've found different scenarios [working on dataset viber](https://github.com/davidberenstein1957/dataset-viber/blob/main/src/dataset_viber/_gradio/_mixins/_import_export.py#L277-L297).

burtenshaw replied:
In the situation where the `Features` contain an `Image` but the values are not a pil the conversiont [function](https://github.com/argilla-io/argilla/blob/8917f5fc8af7b339426d137bd9ff8957c345cb2a/argilla/src/argilla/media.py#L23) with raise an error, so the user will need to cast the dataset themselves.

davidberenstein1957 replied:
@burtenshaw, don't we lose a lot of potential conversion cases only allowing for `PIL`? 

Also, I don't know what happens with the `Image` being in `features` when you load a dataset with `set decode=False` for example. According to the documentation, it still seems to be an `Image` feature but then containing this nested dict or just direct URL, which we could use to ease HF datasets upload IMO, given the code I shared above.

Additionally, I am not sure if `Image` might not also just contain a direct URL.

burtenshaw replied:
The datasets package converts the values with an `Image` `Feature` to PIL image see docs here: https://huggingface.co/docs/datasets/en/about_dataset_features#image-feature

`decode=False` is a parameter of the `Image` class which is passed to `cast_column` of `cast`, It tells `datasets` not to decode the value as a PIL object. I think we can say that we don't support images when the user casts the column with `decode=False`

>it still seems to be an Image feature but then containing this nested dict or just direct URL, which we could use to ease HF datasets upload IMO, given the code I shared above.

Is the direct URL in the object non-ephemeral?

burtenshaw committed:
respond: rename method to _get_image_features

burtenshaw committed:
respond: use list comp

burtenshaw committed:
refactor: simplify map function partially

burtenshaw committed:
respond: take format and mimetype from PIL object

burtenshaw committed:
enhancement: add fallbacks to PNG for unknown image formats

burtenshaw committed:
feat: move image casting into record fields (#5436)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR experiments with cast images within the `Record` class to
abstract away data uri. This supports this flow:

```python
image_paths = [f"{my_image_dir}/{fn}" for fn in os.listdir(my_image_dir)]

records = [
    rg.Record(
        fields={
            "image": image_path,
        },
    )
    for image_path in image_paths
]


dataset.records.log(records)
```

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
